Archive/EulerFermat.lean

```diff
@@ -0,0 +1,169 @@
+
+/-- The odd part of `n > 0` is either 1, or is `∈ [1, n)`. -/
+lemma ord_compl_eq_or_lt (n p : ℕ) (hn : 0 < n) :
+    ord_compl[p] n = 1 ∨ (1 < ord_compl[p] n ∧ n ≥ ord_compl[p] n) := by
+  have h (n m : ℕ) (hn1 : n ≤ m) (hn2 : 1 ≤ n) : n = 1 ∨ (1 < n ∧ m ≥ n) := by omega
+  apply h (ord_compl[p] n)
+  · apply ord_compl_le n
+  · apply ord_compl_pos p (not_eq_zero_of_lt hn)
+
+/-- With respect to prime `p`, the odd part of `p ^ m` is 1 only if `n` is a power of `p`. -/
+lemma ord_compl_of_pow (m n p: ℕ) (hp : p.Prime) (hn : n = p ^ m) : ord_compl[p] n = 1 := by
+  rw [hn, Prime.factorization_pow, Finsupp.single_eq_same]
+  simp only [Prime.pos hp, ofNat_pos, pow_pos, Nat.div_self]
+  exact hp
+
+/-- `n` is a power of `p` only if the odd part of `p ^ m` is 1. -/
+lemma ord_pow_of_compl (n p: ℕ) (hnop : ord_compl[p] n = 1) : ∃ m : ℕ, n = p ^ m := by
+  have h : p ^ n.factorization p * (n / p ^ n.factorization p) = n :=
+    ord_proj_mul_ord_compl_eq_self n p
+  use n.factorization p
+  rw [hnop, mul_one] at h
+  exact h.symm
+
+lemma simplify2 (n : ℕ) (hn : n ≠ 0) :
+    (ord_proj[2] n) * (2 * ((ord_compl[2] n - 1) / 2) + 1) = n :=
+  calc
+    (ord_proj[2] n) * (2 * ((ord_compl[2] n - 1) / 2) + 1) =
+        (ord_proj[2] n) * ((ord_compl[2] n - 1) + 1) := by
+      rw [two_mul_div_two_of_even (n := (ord_compl[2] n - 1))]
+      apply Nat.Odd.sub_odd _ (odd_iff.mpr (by rfl))
+      exact odd_iff.mpr <| two_dvd_ne_zero.mp <| not_dvd_ord_compl prime_two hn
+    _ = (ord_proj[2] n) * (ord_compl[2] n) := by
+      have (n : ℕ) (h : Odd n) : Even (n - 1) := by
+        exact Nat.Odd.sub_odd h (odd_iff.mpr (by rfl))
+      rw [Nat.sub_add_cancel <| zero_lt_of_lt <| ord_compl_pos 2 hn]
+    _ = n := by
+      rw [ord_proj_mul_ord_compl_eq_self]
+
+/-- "If any numbers of the form $a^n+1$ are prime, it is necessary, that they be of the
+  form $a^{2^m}+1$" [Euler, E026] -/
+theorem H3 (a n : ℕ) (ha : 1 < a) (hn : 1 < n)
+    (hP : Nat.Prime (a ^ n + 1)) : ∃ m : ℕ, n = 2 ^ m := by
+  /- First we show that the goal `n = 2 ^ m` is equivalent to the statement
+     "the odd part of `n` is 1", which is written in Lean using the `factorization` function. -/
+  apply ord_pow_of_compl n 2
+  /- The goal is now to show that the odd part of `n` must be 1.
+     The odd part of any number is either 1 (which is equivalent to being a power
+     of two), or greater than one and less or equal `n`. This is hypothesis `h₁`
+     which is proved by applying lemma `ord_compl_eq_or_lt`. -/
+  have h₁ : ord_compl[2] n = 1 ∨ (1 < ord_compl[2] n ∧ n ≥ ord_compl[2] n) :=
+    ord_compl_eq_or_lt n 2 (Nat.zero_lt_of_lt hn)
+  /- If we can disprove the second proposition, then the first remains true in `h₁`,
+     and `h₁` can simply be applied. For this logical argument we need the
+     `true_or_false` lemma, which we apply to `h₁`. -/
+  have true_or_false (A B : Prop) (hB : ¬ B) (hab : A ∨ B) : A := by simp_all only [or_false]
+  apply true_or_false at h₁
```

Comment:
You may be looking for `Or.resolve_right`

Reply:
This changes the logic, so I'd rather leave it. Is it a problem?